internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,142 @@ import (
+type minerInvocationContext interface {
 	Verifier() verification.Verifier
 }
 
+// ActorStateHandler handles the actor state, allowing actors to lock on the state.
+type ActorStateHandler interface {
+	// Take returns the `Cid` for the state and locks the state.
+	//
+	// Any futre calls to `Take` by other instances will block until `Release` or `UpdateRelease` is called.
+	// Review: for @spec, the spec impl of `Take` is not doing any blocking, just panic'ing on the second call on the SAME object instance (this objects are not kept around, so it is always a new instance..).
+	Take() cid.Cid
+	// UpdateRelease updates the actor state and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	UpdateRelease(newStateCID ActorSubstateCID)
+	// Release asserts that the state has not changed and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	// Review: for @spec, why is this method required?
+	Release(checkStateCID ActorSubstateCID)
+}
+
+// IpldStorage2 defines the storage module exposed to actors.
+type IpldStorage2 interface {
+	Put(interface{}) (cid.Cid, error)
+	Get(cid.Cid, interface{}) error
+}
+
+// CallerPattern checks if the caller matches the pattern.
+type CallerPattern interface {
+	IsMatch() bool
+}
+
+// Epoch is the current chain epoch.
+// Dragons: this is just to match names on the VM, Height got renamed to Epoch on the spec
+// TODO: rename types.BlockHeight to block.Epoch (issue: ???)
+type Epoch types.BlockHeight
+
+// Randomness is a string of random bytes
+// Note: placeholder for matching newtypes with the spec.
+type Randomness []byte
+
+// InvocInput are the params to invoke a method in an Actor.
+type InvocInput struct {
+    To      address.Address
+    Method  types.MethodID
+    Params  MethodParams
+    Value   types.AttoFIL
+}
+
+// MethodParam is the parameter to an Actor method already encoded for execution in the VM.
+type MethodParam []byte
+// MethodParams is a list of `MethodParam` to be passed into an Actor method.
+type MethodParams []MethodParam
+
+// Abort will stop the VM execution and return an abort error to the caller.
+// Review: @go-filecoin, do we want to handle this with panics?
+func Abort(msg string) {
+	panic("byte me")
+}
+
+// Assert will abort if the condition is `False` and return an abort error to the caller.
+func Assert(cond bool) {
+	if !cond {
+		Abort("assertion failure in actor code.")
+	}
+}
+
+//
+// Dragons: this will get deleted, still here while we finish some discussions on the PR.
+//
+
+// Runtime defines the ABI interface exposed to actors.
+type Runtime interface {
+	// Review: this is not in the spec ~anymore~ yet..
+	// Note: @jzimmerman: still a WIP, things like the miner actor might need it
```

Comment:
Please prefer to expose the individual values that are needed, if at all possible @jzimmerman

Reply:
Except for a single call I think we are safe to remove.

All calls to `.Message()` access the `.From()` and `.Value()`, except for `paymentbroker.go` which once, access `.CallSeqNum()` to build a `ChannelID`.

The `From`, and `Value` calls can be replaced with the `Caller()` and `ValueReceived()` in the new context.

I'm not certain about the `CallSeqNum()` or how the new actor code will determine the ChannelID without that info being available (randomness perhaps?) @jzimmerman?.

Reply:
There's no payment broker in the spec, go-filecoin has an implementation from a prior spec. In the new world, each channel is a distinct actor.

Reply:
I don't see an issue with exposing `CallSeqNum` to actor code, but we should make clear that this is a sequence number across _top-level_ calls, and isn't incremented by actor method subcalls. (If we need a sequence number with the latter definition, which may be useful in some cases, we should add that separately and make the terminology very clear.)

Reply:
Let's wait for the actor pseudocode to land and see what we need.